File: mio_taskhub/doc_lifecycle.py
# -*- coding: utf-8 -*-
"""文档生命周期状态机（按用户定义的 8 类文档角色表落地）。

| 类型      | 作用       | 生命周期                             |
|---------|----------|------------------------------------|
| PRD     | 为什么做    | Draft → Approved                   |
| Spec    | 怎么设计    | Draft → Review → Approved          |
| 接口契约  | 怎么调用    | Draft → Review → Approved          |
| ADR     | 为什么这么决定 | Proposed → Accepted → Superseded   |
| Plan    | 怎么实施    | Draft → Approved → Done            |
| Task    | 谁做什么    | Todo → Doing → Done（即任务本身 state 机，非文档）|
| Test    | 是否正确    | Planned → Passed/Failed            |
| Release | 发布了什么   | Planned → Released                 |
| Incident| 出了什么问题  | Open → Resolved → Closed           |

扩展（2026-09-17）：**接口契约 `api`** 与 Spec 同形（`draft → review → approved`）。
用户要求接口契约「事无巨细」，其质量规格是文档链里最严的一份；把它纳入生命周期后，
`set_doc_status` 的质量门控才会作用于它——否则质量 error 仅出现在写入响应与报告里、
不阻断任何推进（这是此前的缺口）。

kind 映射（复用 21+1 类）：PRD→requirement、ADR→decision、Release→milestone、
Incident→incident（新增 kind）、接口契约→api。未列入表的 kind 无生命周期，不支持设状态。
状态存 `task.doc_statuses`（kind -> {state, at, note}），转移事件走 events。
"""

# kind -> 状态机定义；transitions 只允许向前（严格线性，回退走重写文档重新起草）
DOC_LIFECYCLE = {
    'requirement': {   # PRD
        'states': ('draft', 'approved'),
        'transitions': {'draft': ('approved',)},
    },
    'spec': {
        'states': ('draft', 'review', 'approved'),
        'transitions': {'draft': ('review',), 'review': ('approved',)},
    },
    'api': {           # 接口契约（与 spec 同形；质量规格最严，见 doc_quality）
        'states': ('draft', 'review', 'approved'),
        'transitions': {'draft': ('review',), 'review': ('approved',)},
    },
    'decision': {      # ADR
        'states': ('proposed', 'accepted', 'superseded'),
        'transitions': {'proposed': ('accepted',), 'accepted': ('superseded',)},
    },
    'plan': {
        'states': ('draft', 'approved', 'done'),
        'transitions': {'draft': ('approved',), 'approved': ('done',)},
    },
    'test': {
        'states': ('planned', 'passed', 'failed'),
        'transitions': {'planned': ('passed', 'failed')},
    },
    'milestone': {     # Release
        'states': ('planned', 'released'),
        'transitions': {'planned': ('released',)},
    },
    'incident': {
        'states': ('open', 'resolved', 'closed'),
        'transitions': {'open': ('resolved',), 'resolved': ('closed',)},
    },
}

# kind -> 初始状态（文档首次写入时自动落位）
INITIAL_STATE = {
    'requirement': 'draft',
    'spec': 'draft',
    'api': 'draft',
    'decision': 'proposed',
    'plan': 'draft',
    'test': 'planned',
    'milestone': 'planned',
    'incident': 'open',
}


def lifecycle_of(kind):
    """返回该 kind 的生命周期定义；无生命周期的 kind 返回 None。"""
    return DOC_LIFECYCLE.get(kind)


def initial_state_of(kind):
    """该 kind 的初始状态；无生命周期返回 None。"""
    return INITIAL_STATE.get(kind)


def allowed_next(kind, state):
    """当前状态的合法后继状态列表（无生命周期/终态 → 空列表）。"""
    lc = DOC_LIFECYCLE.get(kind)
    if not lc or state not in lc['transitions']:
        return []
    return list(lc['transitions'][state])


def validate_transition(kind, current, target):
    """校验状态转移。返回 None 表示合法，否则返回错误文案。"""
    lc = DOC_LIFECYCLE.get(kind)
    if not lc:
        return f"kind '{kind}' has no document lifecycle"
    if target not in lc['states']:
        return (f"invalid state '{target}' for kind '{kind}'; "
                f"states: {', '.join(lc['states'])}")
    if current is None:
        if target != INITIAL_STATE.get(kind):
            return (f"first status of '{kind}' must be "
                    f"'{INITIAL_STATE.get(kind)}', got '{target}'")
        return None
    if current == target:
        return f"kind '{kind}' is already '{current}'"
    if target not in allowed_next(kind, current):
        return (f"illegal transition {current} -> {target} for kind '{kind}'; "
                f"allowed: {', '.join(allowed_next(kind, current)) or 'none (terminal)'}")
    return None

File: web/src/docLifecycle.js
// 文档生命周期状态的展示映射。
//
// 与后端 mio_taskhub/doc_lifecycle.py 的 DOC_LIFECYCLE 保持一致（8 类文档）。
// 后端状态存 task.doc_statuses[kind] = { state, at, note }，清单接口
// GET /tasks/{id}/documents 的每条条目带 status 字段；总览接口
// GET /tasks/{id}/doc/statuses 额外给出 states 全序列与 allowed_next。
//
// 这里只做「状态 → 中文名 + 色调」的纯映射，不含副作用，供 DocPanel 等复用。
// 词表变更时请同步本文件。

// 状态 → 中文名（跨 kind 共用同一套词）
export const DOC_STATE_LABELS = {
  draft: '草稿',
  review: '待审',
  approved: '已批准',
  proposed: '提议',
  accepted: '已接受',
  superseded: '已取代',
  planned: '已计划',
  done: '已完成',
  passed: '通过',
  failed: '未通过',
  released: '已发布',
  open: '待处理',
  resolved: '已解决',
  closed: '已关闭',
}

// 状态 → 色调：
//   pending   起点/未开始推进
//   attention 待人工审核（卡在这里需要有人/agent 动作）
//   positive  目标达成
//   danger    负向终态（需处理）
//   neutral   归档态（不再变化，但不是失败）
const DOC_STATE_TONES = {
  draft: 'pending',
  proposed: 'pending',
  planned: 'pending',
  open: 'pending',
  review: 'attention',
  approved: 'positive',
  accepted: 'positive',
  done: 'positive',
  passed: 'positive',
  released: 'positive',
  resolved: 'positive',
  failed: 'danger',
  superseded: 'neutral',
  closed: 'neutral',
}

export const docStateLabel = (state) =>
  state ? (DOC_STATE_LABELS[state] || state) : '未落状态'

export const docStateTone = (state) => DOC_STATE_TONES[state] || 'pending'

// 清单条目 / doc_statuses 条目 → 状态字符串（两者形状不同，统一在这里取）
//   清单条目:   { kind, status: { state, at, note } | null }
//   statuses:   { state, at, note, states, allowed_next, has_doc }
export const stateOfDoc = (doc) => (doc && doc.status && doc.status.state) || null

export const stateOfStatus = (entry) => (entry && entry.state) || null

// 生命周期推进条：把 states 全序列标注为 done / current / todo
// 返回 [{ state, label, tone, phase }]，phase ∈ done | current | todo
export function lifecycleTrack(states, current) {
  const list = Array.isArray(states) ? states : []
  const idx = current ? list.indexOf(current) : -1
  return list.map((s, i) => ({
    state: s,
    label: docStateLabel(s),
    tone: docStateTone(s),
    phase: idx === -1 ? 'todo' : i < idx ? 'done' : i === idx ? 'current' : 'todo',
  }))
}
